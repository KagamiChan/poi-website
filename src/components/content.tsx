import classnames from 'classnames'
import { Dropdown, IDropdownOption } from 'office-ui-fabric-react/lib/Dropdown'
import React, { useMemo, useState } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'
import { UAParser } from 'ua-parser-js'

import { DownloadCards } from './download-cards'
import { TypeCat } from './type-cat'

import poi from '../assets/poi.png'
import styles from './content.module.css'

export const getSettings = () => {
  const { os, cpu } = new UAParser().getResult()
  let initialSlide = 0
  if (os.name === 'Linux') {
    initialSlide = 0
  } else if (os.name === 'Debian' || os.name === 'Ubuntu') {
    initialSlide = 1
  } else if (os.name === 'CentOS' || os.name === 'Fedora') {
    initialSlide = 2
  } else if (os.name === 'Mac OS') {
    initialSlide = 3
  } else if (os.name === 'Windows') {
    initialSlide = 5
    if (cpu.architecture === 'ia64' || cpu.architecture === 'amd64') {
      initialSlide = 7
    }
  }
  return {
    infinite: true,
    initialSlide,
    slidesToScroll: 1,
    slidesToShow: 1,
    speed: 500,
  }
}

const getTargetIndex = () => {
  const { os, cpu } = new UAParser().getResult()
  if (os.name === 'Linux') {
    return 0
  } else if (os.name === 'Debian' || os.name === 'Ubuntu') {
    return 1
  } else if (os.name === 'CentOS' || os.name === 'Fedora') {
    return 2
  } else if (os.name === 'Mac OS') {
    return 3
  } else if (os.name === 'Windows') {
    if (cpu.architecture === 'ia64' || cpu.architecture === 'amd64') {
      return 7
    }
    return 5
  }
  return 0
}

const version = {
  beta: 'v10.4.0',
  betaAvailable: true,
  stable: 'v10.3.0',
}

const targets = [
  'linux-x64',
  'linux-x64',
  'linux-rpm-x64',
  'macos-x64',
  'win-ia32',
  'win-ia32-setup',
  'win-x64',
  'win-x64-setup',
]

const StyledDropdown = styled(Dropdown)`
  width: 20em;
`

const CenterContainer = styled.div`
  display: flex;
  align-items: center;
  margin: 0 auto;
`

const Description = styled.div`
  font-size: 36px;
  text-align: center;
  white-space: pre-line;

  @media screen and (min-width: 768px) {
    font-size: 48px;
  }
`

export const Content = () => {
  const { t } = useTranslation()

  const options: IDropdownOption[] = useMemo(
    () =>
      targets.map(target => ({
        key: target,
        text: t(target),
      })),
    [t],
  )

  const [selected, setSelected] = useState(targets[getTargetIndex()])

  return (
    <div className={styles.container}>
      <div className={styles.title}>
        <img
          src={poi}
          className={classnames(styles.logo, {
            [styles.aprilfoolsday]:
              new Date().getMonth() === 3 && new Date().getDate() === 1,
          })}
        />
        <span className={styles.name}>{t('name')}</span>
      </div>
      <Description>
        <TypeCat text={t('description')} />
      </Description>
      <DownloadCards target={selected} version={version} />
      <CenterContainer>
        {t('Choose another platform')}
        <StyledDropdown
          options={options}
          selectedKey={selected}
          onChange={(_, item) => setSelected(item!.key as string)}
        />
      </CenterContainer>
      <CenterContainer>
        <a href="https://npm.taobao.org/mirrors/poi" target="_blank">
          {t('other-versions')}
        </a>
      </CenterContainer>
    </div>
  )
}
