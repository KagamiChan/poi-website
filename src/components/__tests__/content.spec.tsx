import { fireEvent, waitForElement } from '@testing-library/react'
import React from 'react'
import { renderWithTheme } from '../../testing-utils'

import { Content } from '../content'

jest.mock('ua-parser-js')

describe('<Content /> with beta > stable', () => {
  let fetch: jest.SpyInstance
  beforeAll(() => {
    fetch = jest.spyOn(window, 'fetch').mockImplementation(
      async () =>
        new Response(
          JSON.stringify({
            betaVersion: 'v10.5.0-beta.0',
            version: 'v10.4.0',
          }),
        ),
    )
  })

  afterAll(() => {
    fetch.mockReset()
  })

  it('renders', async () => {
    const { asFragment, getByTestId } = renderWithTheme(<Content />)
    await waitForElement(() => getByTestId('download-stable-version'))
    expect(asFragment()).toMatchSnapshot()
  })

  it('opens target list', async () => {
    const { asFragment, getByTestId } = renderWithTheme(<Content />)
    await waitForElement(() => getByTestId('download-stable-version'))
    fireEvent.click(getByTestId('open-dialog'))
    await waitForElement(() => getByTestId('modal-container'))
    expect(asFragment()).toMatchSnapshot()
  })
})

describe('<Content /> with beta < stable', () => {
  let fetch: jest.SpyInstance
  beforeAll(() => {
    fetch = jest.spyOn(window, 'fetch').mockImplementation(
      async () =>
        new Response(
          JSON.stringify({
            betaVersion: 'v10.3.0-beta.0',
            version: 'v10.4.0',
          }),
        ),
    )
  })

  afterAll(() => {
    fetch.mockReset()
  })

  it('renders', async () => {
    const { asFragment, getByTestId } = renderWithTheme(<Content />)
    await waitForElement(() => getByTestId('download-stable-version'))
    expect(asFragment()).toMatchSnapshot()
  })

  it('opens target list', async () => {
    const { asFragment, getByTestId } = renderWithTheme(<Content />)
    await waitForElement(() => getByTestId('download-stable-version'))
    fireEvent.click(getByTestId('open-dialog'))
    await waitForElement(() => getByTestId('modal-container'))
    expect(asFragment()).toMatchSnapshot()
  })
})
