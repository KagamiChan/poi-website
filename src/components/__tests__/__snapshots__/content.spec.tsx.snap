// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Content /> renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bZQynM gSJnjk"
  >
    <h1
      class="sc-gzVnrw eJmVhY"
    >
      <img
        alt="poi"
        class="sc-htoDjs jvwLIY"
        src="poi.png"
      />
      <span
        class="sc-dnqmqq bcJpYV"
      >
        poi
      </span>
    </h1>
    <div
      class="sc-gqjmRU iaGhZ"
    >
      <div
        class="sc-ifAKCX kWQbT"
      >
        <div
          class="sc-EHOje gYoJOI"
        >
          <div
            data-testid="typist"
          />
        </div>
        <div
          class="sc-EHOje dZEEWs"
          data-testid="typing"
        >
          Scalable KanColle browser and tool.
          <span
            style="opacity: 0;"
          >
            _
          </span>
        </div>
      </div>
    </div>
    <div
      class="sc-bdVaJa gbaexa"
    >
      <div
        class="sc-bwzfXH gFmFes"
      >
        Download for Windows (x64)
      </div>
      <a
        data-testid="download-stable-version"
        href="https://npm.taobao.org/mirrors/poi/v10.4.0/poi-setup-10.4.0.exe"
      >
        <button
          class="sc-htpNat kEBVTB"
        >
          <div>
            v10.4.0
          </div>
          <div
            class="sc-bxivhb gEfMvB"
          >
            Stable, for most users
          </div>
        </button>
      </a>
    </div>
    <div
      class="sc-gZMcBi hlXFsd"
    >
      <div
        class="ms-Dropdown-container sc-iwsKbI jSRktV root-50"
      >
        <label
          class="ms-Label ms-Dropdown-label label-68"
          for="Dropdown1"
          id="Dropdown1-label"
        >
          Choose another platform
        </label>
        <div
          aria-describedby="Dropdown1-option"
          aria-expanded="false"
          aria-labelledby="Dropdown1-label"
          class="ms-Dropdown dropdown-52"
          data-is-focusable="true"
          id="Dropdown1"
          role="listbox"
          tabindex="0"
        >
          <span
            aria-atomic="true"
            aria-label="Windows (x64)"
            aria-live="off"
            aria-posinset="8"
            aria-selected="true"
            aria-setsize="8"
            class="ms-Dropdown-title title-53"
            id="Dropdown1-option"
            role="option"
          >
            <span>
              Windows (x64)
            </span>
          </span>
          <span
            class="ms-Dropdown-caretDownWrapper caretDownWrapper-54"
          >
            <i
              aria-hidden="true"
              class="ms-Dropdown-caretDown caretDown-69"
              data-icon-name="ChevronDown"
              role="presentation"
            />
          </span>
        </div>
      </div>
    </div>
    <div
      class="sc-gZMcBi hlXFsd"
    >
      <a
        class="sc-VigVT ecCZLv"
        href="https://npm.taobao.org/mirrors/poi"
        rel="noopener noreferrer"
        target="_blank"
      >
        Check the complete list 
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-external-link-alt fa-w-18 "
          data-icon="external-link-alt"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M576 24v127.984c0 21.461-25.96 31.98-40.971 16.971l-35.707-35.709-243.523 243.523c-9.373 9.373-24.568 9.373-33.941 0l-22.627-22.627c-9.373-9.373-9.373-24.569 0-33.941L442.756 76.676l-35.703-35.705C391.982 25.9 402.656 0 424.024 0H552c13.255 0 24 10.745 24 24zM407.029 270.794l-16 16A23.999 23.999 0 0 0 384 303.765V448H64V128h264a24.003 24.003 0 0 0 16.97-7.029l16-16C376.089 89.851 365.381 64 344 64H48C21.49 64 0 85.49 0 112v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V287.764c0-21.382-25.852-32.09-40.971-16.97z"
            fill="currentColor"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
